aiida_quantumespresso/calculations/pp.py

```diff
@@ -1,27 +1,218 @@
+
+    def prepare_for_submission(self, folder): # pylint: disable=too-many-branches,too-many-statements
+        """Create the input files from the input nodes passed to this instance of the `CalcJob`.
+
+        :param folder: an `aiida.common.folders.Folder` to temporarily write files on disk
+        :return: `aiida.common.datastructures.CalcInfo` instance
+        """
+
+        # Put the first-level keys as uppercase (i.e., namelist and card names) and the second-level keys as lowercase
+        if 'parameters' in self.inputs:
+            parameters = _uppercase_dict(self.inputs.parameters.get_dict(), dict_name='parameters')
+            parameters = {k: _lowercase_dict(v, dict_name=k) for k, v in six.iteritems(parameters)}
+        else:
+            parameters = {}
+
+        # Set default values. NOTE: this is different from PW/CP
+        for blocked in self._blocked_keywords:
+            namelist = blocked[0].upper()
+            key = blocked[1].lower()
+            value = blocked[2]
+
+            if namelist in parameters:
+                if key in parameters[namelist]:
+                    raise exceptions.InputValidationError(
+                        "You cannot specify explicitly the '{}' key in the '{}' "
+                        'namelist.'.format(key, namelist))
+            else:
+                parameters[namelist] = {}
+            parameters[namelist][key] = value
+
+        # Check for essential keys
+        try:
+            plot_num = parameters['INPUTPP']['plot_num']
+            dimension = parameters['PLOT']['iflag']
+        except KeyError:
+            raise exceptions.InputValidationError("Both 'plot_num' and 'iflag' must be set")
+
+        # Check that a valid plot type is requested
+        if plot_num in range(23) and plot_num not in [14, 15, 16]: # Must be integer in range 0-22, but not 14-16:
+            parameters['INPUTPP']['plot_num'] = int(plot_num) # If this test passes, we can safely cast to int
+        else:
+            raise exceptions.InputValidationError("'plot_num' must be an integer in the range 0-23")
+
+        # Check for valid plot dimension:
+        if dimension in range(5): # Must be in range 0-4:
+            parameters['PLOT']['iflag'] = int(dimension)
+        else:
+            raise exceptions.InputValidationError("'iflag' must be an integer in the range 0-4")
+
+        # Restrict the plot output to the file types that we want to be able to parse
+        if dimension in [0, 1, 4]:
+            parameters['PLOT']['output_format'] = 0    # Gnuplot, 1D
+        elif dimension == 2:
+            parameters['PLOT']['output_format'] = 7    # Gnuplot, 2D
+        elif dimension == 3:
+            parameters['PLOT']['output_format'] = 6    # Gaussian cube, 3D
+        else:
+            raise exceptions.InputValidationError(
+                "Unsupported value, {}, for 'iflag'. Must be in the range 0 to 4".format(dimension)
+            )
+
```

Comment:
These validation rules can be combined, no?

Reply:
Yes, combined into a lookup dict to set `output_format`. The actual validation of `iflag` was already done.